File: Python/math/Next_Prime_Number.py
'''
INTRODUCTION

The code below calculates and returns the next prime number from the number given by user.
It takes a number n from user and returns a number x such that
1. x is greater than n
2. x is a prime number and 
3. there are no prime numbers between n and x 

'''

def next_prime(n):
	# looping from next number to find if it prime or not
	x = n+1
	while True:
		flag = False
		# checking whether x is divisible by numbers from 2 to it's half
		for y in range(2,x//2+1):
			if x%y==0:
				x+=1
				flag = True
				break
		if flag==False:
			return(x)

n = int(input("Enter a number : "))
print('Next prime number is ', next_prime(n))

'''
OUTPUT

Enter a number : 23
Next prime number is  29


TIME COMPLEXITY : O(n²)
SPACE COMPLEXITY : O(1)
'''

File: C-Plus-Plus/math/next_prime_number.cpp
/*
The next prime number p of a number n is a number such that
- p is greater than n.
- p is a prime number.
- There are no prime numbers between n and p.
*/

#include<bits/stdc++.h>
using namespace std;

int next_prime(int n){
    int p = n + 1;
    if(p<=1){
        return 2;
    }
    while(true){
        //consider that p is not a composite number
        bool flag = false;
        for(int i=2; i<=floor(n/2)+1; i++){
            //found a factor other than 1 or p
            if(p%i==0){
                p = p + 1;
                //p is composite
                flag = true;
                break;
            }
        }
        //p is prime
        if(flag==false){
            return p;
        }
    }
}

int main(){
    int n;
    cout<<"Enter a number : ";
    cin>>n;
    cout<<"Next prime number is "<<next_prime(n);
}

/* Sample I/O:
Enter a number : 55
Next prime number is 59

Time Complexity : O(n^2)
Space Complexity : O(1)
*/
